tree/tree/inc/ROOT/InternalTreeUtils.hxx

```diff
@@ -61,18 +61,25 @@ struct RFriendInfo {
    */
    std::vector<std::vector<std::string>> fFriendChainSubNames;
 
-   void AddFriend(const std::string &treeName, const std::string &fileNameGlob, const std::string &alias = "");
+   /**
+    * Entries of each tree in each friend.
+    */
+   std::vector<std::vector<Long64_t>> fFriendEntries;
+
+   void AddFriend(const std::string &treeName, const std::string &fileNameGlob, const std::string &alias = "",
+                  Long64_t treeEntries = TTree::kMaxEntries);
 
-   void
-   AddFriend(const std::string &treeName, const std::vector<std::string> &fileNameGlobs, const std::string &alias = "");
+   void AddFriend(const std::string &treeName, const std::vector<std::string> &fileNameGlobs,
+                  const std::string &alias = "", const std::vector<Long64_t> &treeEntries = {});
 
    void AddFriend(const std::vector<std::pair<std::string, std::string>> &treeAndFileNameGlobs,
-                  const std::string &alias = "");
+                  const std::string &alias = "", const std::vector<Long64_t> &treeEntries = {});
 };
 
 std::vector<std::string> GetTopLevelBranchNames(TTree &t);
 std::vector<std::string> GetFileNamesFromTree(const TTree &tree);
-RFriendInfo GetFriendInfo(const TTree &tree);
+RFriendInfo GetFriendInfo(const TTree &tree, bool retrieveEntries = false);
+std::vector<std::unique_ptr<TChain>> MakeFriendsAndAddToTree(TTree &tree, const RFriendInfo &finfo);
```

Comment:
I'm not a _huge_ fan of this function because it's doing 2 separate things (as the name suggests). More importantly it both acts on the `tree` variable _and_ has a return value that shouldn't be discarded. It's easy to use it as `MakeFriendsAndAddToTree(tree, finfo)`, let the return value go out of scope and cause trouble.

This is not much harder to read and much harder to get wrong:
```cpp
const auto friends = MakeFriends(finfo);
for (auto &friend : friends)
  tree->AddFriend(friend.get())
```

Reply:
Fair enough, the full logic should also deal with friend aliases. In the next commit this has thus been reworked to:
```cpp
   fFriends = ROOT::Internal::TreeUtils::MakeFriends(friendInfo);
   const auto nFriends = friendInfo.fFriendNames.size();
   R__ASSERT(nFriends == fFriends.size() && "Created the wrong number of friends from the available information.");
   for (std::size_t i = 0ul; i < nFriends; i++){
      const auto &thisFriendAlias = friendInfo.fFriendNames[i].second;
      fChain->AddFriend(fFriends[i].get(), thisFriendAlias.c_str());
   }
```